openapi3/internalize_refs.go

```diff
@@ -0,0 +1,373 @@
+package openapi3
+
+import (
+	"context"
+	"strconv"
+	"strings"
+)
+
+func schemaNames(s Schemas) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func parametersMapNames(s ParametersMap) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+// InternalizeRefs removes all references to external files from the spec and moves them
+// to the components section.
+//
+// refNameResolver takes in references to returns a name to store the reference under locally.
+// Some care should be taken to make sure the resolver does not return duplicate names for different
+// references.
+func (spec *T) InternalizeRefs(ctx context.Context, refNameResolver func(ref string) string) *T {
```

Comment:
* Please use `doc` identifier instead of `spec` (as is done throughout the source)
* Any reason to output `*T`? This seems redundant as the receiver is a pointer.
* `error` should be outputted, instead of `panic`ing on loops.
* `refNameResolver` should have a default implementation (ie. the one given in tests)

Reply:
- No problem!
- Habit I am in for my APIs to indicate you are mutating the state, but yes its redundant I'll remove it
- As later noted I will just remove it
- Easy!